synapse/handlers/message.py

```diff
@@ -949,14 +950,24 @@ async def create_new_client_event(
         else:
             prev_event_ids = await self.store.get_prev_events_for_room(builder.room_id)
 
-        # we now ought to have some prev_events (unless it's a create event).
-        #
-        # do a quick sanity check here, rather than waiting until we've created the
+        # Do a quick sanity check here, rather than waiting until we've created the
         # event and then try to auth it (which fails with a somewhat confusing "No
         # create event in auth events")
-        assert (
-            builder.type == EventTypes.Create or len(prev_event_ids) > 0
-        ), "Attempting to create an event with no prev_events"
+        room_version_obj = await self.store.get_room_version(builder.room_id)
+        if room_version_obj.msc2716_empty_prev_events:
+            # We allow events with no `prev_events` but it better have some `auth_events`
+            assert (
+                builder.type == EventTypes.Create
+                or len(prev_event_ids) > 0
+                # Allow an event to have empty list of prev_event_ids
+                # only if it has auth_event_ids.
+                or (auth_event_ids and len(auth_event_ids) > 0)
```

Comment:
I'd be tempted to add an assertion that this is an insertion event, to ensure that we're doing the right thing elsewhere. Specifically, we do have a bug somewhere that means that we sometimes try and send events without any prev events and it would be good to keep catching those.

Reply:
```suggestion
                or auth_event_ids
```

if `len(auth_event_ids) == 0` then `auth_event_ids` will be falsey, so the extra check is redundant.

Reply:
> I'm not enthusiastic about any change which makes it more likely for us to start accidentally creating regular events without prev_events, so +1 to the suggestion that we only allow this for insertion events.

If we're only allowing for certain type of events, we need to allow no `prev_events` for all the state events too. The state event chain (the alice events at the top) is actually the one with no `prev_events`. We connect the insertion event to the floating state chain.

The only time, the insertion event itself is floating is when no `state_events_at_start` are defined.

![](https://user-images.githubusercontent.com/558581/140976606-1f8da184-c2b8-4035-abec-a2668e052350.png)

Reply:
> Is it really a problem to have events without `prev_events` though? They seem like they would be benign and a much better alternative to a homeserver lying out their teeth about a fake `prev_event`.

The problem is that for the normal case not having any `prev_event` means that we won't be able to calculate the state at the event.

---

One thought I had: can we allow events with no prev events if and only if they're either a) an insertion event or b) an outlier?

Separately: I'm not sure its correct that the insertion event points to anything else, as we need want federation to ask for the state at the insertion event, rather than any of the state events that it references?

Reply:
> The problem is that for the normal case not having any `prev_event` means that we won't be able to calculate the state at the event.

Perhaps there are two separate concerns here:

 1. Just making sure app code is doing the right thing for the majority of cases. We expect `prev_events` in all of the non-historical cases.
 1. Having events with no `prev_events` being a problem in a room


For 1, we can add a `allow_no_prev_events: [True|False]` parameter to `create_new_client_event` to only allow it when specified.

For 2, see https://github.com/matrix-org/synapse/pull/11243#discussion_r750816377. I think events with no `prev_events` work just fine in a room. If having no `prev_events` is a problem, then we probably shouldn't do it for the historical events either. Things seem to work without needing to worry about anything.

> One thought I had: can we allow events with no prev events if and only if they're either a) an insertion event or b) an outlier?

I think that would work but I'm leaning away from making any special cases.

Reply:
@erikjohnston What are your thoughts here?

Reply:
(I swear this thread wasn't here yesterday, sorry :confused: )

> For 1, we can add a allow_no_prev_events: [True|False] parameter to create_new_client_event to only allow it when specified.

I think that's fine. I understand wanting to avoid special casing, but my assumption was that we'd need to add special handling for events with no prev events anyway, so it'd make sense to reuse that.

Reply:
Removed the new room version and added `allow_no_prev_events: bool = False` to only allow this case whenever we explicitly set it ✅